Documentation/Plugin-developer-manual.md

```diff
@@ -0,0 +1,230 @@
+
+
+```java
+uint64_t getCapabilities(void)
+```
+
+This is a simple getter for router and filter capabilities. The return value is a bitfield resulting from ORring the individual capabilities. `routing.h` lists the allowed capability flags. The most common capability to set is `RCAP_TYPE_STMT_INPUT`, causing the protocol modules (if compatible) to only send complete statements to the filter/router chain.
+
+#### Monitor
+
+```java
+void *(*startMonitor)(MXS_MONITOR *monitor, const MXS_CONFIG_PARAMETER *params);
+void (*stopMonitor)(MXS_MONITOR *monitor);
+void (*diagnostics)(DCB *, const MXS_MONITOR *);
+```
+
+Monitor modules typically run a repeated monitor routine with a used defined interval. The `MXS_MONITOR` is a standard monitor definition used for all monitors and contains a void pointer for storing module specific data. `startMonitor` should create a new thread for itself using functions in the CPI and have it regularly run a monitor loop. In the beginning of every monitor loop, the monitor should lock the `SERVER`-structures of its servers. This prevents any administrative action from interfering with the monitor during its pass.
+
+## Core public interface
+
+One could spend pages explaining the public headers in detail, what to do?
+
+## Compiling and installing
+
+The requirements for compiling a module are:
+* The public headers (CPI)
+* A compatible compiler, typically GCC (version X??)
+* Libraries required by the public headers
+
+The public header files themselves may include headers from other libraries. These libraries need to be installed and it may be required to point out their location to gcc. Some of the more commonly required libraries are:
+ * *mySQL Connector-C* (mySQL or MariaDB version libmariadb-client-lgpl-dev), used by the mySQL protocol module
+ * *pcre2 regular expressions* (libpcre2-dev), used for example by the header `modutil.h`
+
+After all dependencies are accounted for, the module should compile with a command along the lines of
+```
+gcc -Iinclude/ -I/usr/include/mariadb -shared -fPIC -g -o libmymodule.so mymodule.cpp
+```
+For large modules one may need to set up more complicated compilation, but that is not in the scope of this document.
+
+
+## Running, testing and profiling?
+
+Is this section needed?
```

Comment:
A complete step by step procedure for building a module, installing it in MaxScale, starting MaxScale, seeing how the module logs some startup noise, connecting to MaxScale via maxadmin and via maxadmin causing the module to print "Hello, World!" would be informative.